src/merkle/smt/mod.rs

```diff
@@ -140,6 +140,63 @@ pub(crate) trait SparseMerkleTree<const DEPTH: u8> {
         self.set_root(node_hash);
     }
 
+    /// Start a prospective mutation transaction, which can be queried, discarded, or applied.
+    ///
+    /// This method returns a [`MutationSet`], on which you can call [`MutationSet::insert()`] to
+    /// perform prospective mutations to this Merkle tree, and check the computed root hash given
+    /// those mutations with [`MutationSet::root()`]. When you are done, call
+    /// [`MutationSet::done()`], and pass that to [`Smt::commit()`] to apply the prospective
+    /// mutations to this tree. Or, to discard the chanes, simply [`drop`] the [`MutationSet`].
+    ///
+    /// ## Example
+    /// ```
+    /// # use miden_crypto::{hash::rpo::RpoDigest, Felt};
+    /// # use miden_crypto::merkle::{Smt, EmptySubtreeRoots, SMT_DEPTH};
+    /// # let mut smt = Smt::default();
+    /// let mut mutations = smt.mutate();
+    /// mutations.insert(Default::default(), Default::default());
+    /// assert_eq!(mutations.root(), *EmptySubtreeRoots::entry(SMT_DEPTH, 0));
+    /// smt.commit(mutations.done());
+    /// assert_eq!(smt.root(), *EmptySubtreeRoots::entry(SMT_DEPTH, 0));
+    /// ```
+    fn mutate(&self) -> MutationSet<DEPTH, Self>
+    where
+        Self: Sized,
+    {
+        MutationSet {
+            tree: self,
+            node_mutations: Default::default(),
+            new_pairs: Default::default(),
+            new_root: self.root(),
+        }
+    }
```

Comment:
Question: what is the advantage of making the `MutationSet` a user-manipulatable object? I'm specifically thinking in comparison with the interface suggested in https://github.com/0xPolygonMiden/crypto/issues/222#issuecomment-1824072822. Basically, why not have something like:
```rust
fn compute_mutations(
    &self,
    kv_pairs: impl IntoIter<Item = (Self::Key, Self::Value)>,
) -> MutationSet<Self::Key, Self::Value> {
    ...
}

fn apply_mutations(&mut self, mutations: MutationSet<Self::Key, Self::Value>) {
    ...
}
```
It seems like this approach would be a bit simpler to implement and to use, and shouldn't sacrifice much flexibility - but maybe I'm missing something?

Reply:
There is an advantage in that the root can be checked multiple times before applying a set of changes. This interface is definitely simpler though, so if you don't need that I can very easily change it to this instead.

Reply:
I would probably go with the simpler interface. It also more closely aligns with the use case we have in mind for this. For example, by providing a whole list of updates up-front, we may be able to do some optimizations in the future (e.g., somehow make use of multi-threading).

Reply:
Alright, I'll get right to that.